src/runner_metrics.py

```diff
@@ -0,0 +1,222 @@
+        return None
+
+    logger.debug("Runner %s installed at %s", fs.runner_name, installed_timestamp)
+    try:
+        pre_job_metrics = json.loads(fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME).read_text())
+        logger.debug("Pre-job metrics for runner %s: %s", fs.runner_name, pre_job_metrics)
+    except FileNotFoundError:
+        logger.warning("%s not found for runner %s.", PRE_JOB_METRICS_FILE_NAME, fs.runner_name)
+        return None
+    except JSONDecodeError as exc:
+        raise CorruptMetricDataError(str(exc)) from exc
+
+    try:
+        post_job_metrics = json.loads(fs.path.joinpath(POST_JOB_METRICS_FILE_NAME).read_text())
+        logger.debug("Post-job metrics for runner %s: %s", fs.runner_name, post_job_metrics)
+    except FileNotFoundError:
+        logger.warning("%s not found for runner %s", POST_JOB_METRICS_FILE_NAME, fs.runner_name)
+        post_job_metrics = None
+    except JSONDecodeError as exc:
+        raise CorruptMetricDataError(str(exc)) from exc
+
+    if not isinstance(pre_job_metrics, dict):
+        raise CorruptMetricDataError(
+            f"Pre-job metrics for runner {fs.runner_name} is not a JSON object."
+        )
+
+    if not isinstance(post_job_metrics, dict) and post_job_metrics is not None:
+        raise CorruptMetricDataError(
+            f"Post-job metrics for runner {fs.runner_name} is not a JSON object."
+        )
+
+    try:
+        return RunnerMetrics(
+            installed_timestamp=installed_timestamp,
+            pre_job=PreJobMetrics(**pre_job_metrics),
+            post_job=PostJobMetrics(**post_job_metrics) if post_job_metrics else None,
+        )
+    except ValidationError as exc:
+        raise CorruptMetricDataError(str(exc)) from exc
+
+
+def _issue_runner_metrics(runner_metrics: RunnerMetrics, flavor: str) -> None:
+    """Issue metrics.
+
+    Converts the metrics into respective metric events and issues them.
+
+    Args:
+        runner_metrics: The metrics to be issued.
+        flavor: The flavor of the runners.
+    """
+    event = metrics.RunnerStart(
+        timestamp=runner_metrics.pre_job.timestamp,
+        flavor=flavor,
+        workflow=runner_metrics.pre_job.workflow,
+        repo=runner_metrics.pre_job.repository,
+        github_event=runner_metrics.pre_job.event,
+        idle=runner_metrics.pre_job.timestamp - runner_metrics.installed_timestamp,
+    )
+    metrics.issue_event(event)
+
+
+def _clean_up_shared_fs(fs: shared_fs.SharedFilesystem) -> None:
+    """Clean up the shared filesystem.
+
+    Remove all metric files and afterwards the shared filesystem.
+    Args:
+        fs: The shared filesystem for a specific runner.
+    """
+    fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME).unlink(missing_ok=True)
+    fs.path.joinpath(POST_JOB_METRICS_FILE_NAME).unlink(missing_ok=True)
+    fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME).unlink(missing_ok=True)
```

Comment:
Isn't it the responsability of `shared_fs` to empty the fs directory ? I think it's anyway doing `rmtree` on the folder

Reply:
The idea was to remove the metric files explicitly, and let an error pass through in order to avoid keeping the files and logging the same metric multiple times in the next reconciliation pass.

Reply:
Code has been updated, files are still explicitly unlinked but errors are caught.